lib/_stream_wrap.js

```diff
@@ -27,21 +32,29 @@ function StreamWrap(stream) {
     return self.readStop();
   };
   handle.onshutdown = function(req) {
-    return self.shutdown(req);
+    return self.doShutdown(req);
   };
   handle.onwrite = function(req, bufs) {
-    return self.write(req, bufs);
+    return self.doWrite(req, bufs);
   };
 
   this.stream.pause();
+  this.stream.on('error', function(err) {
+    self.emit('error', err);
+  });
   this.stream.on('data', function(chunk) {
-    self._handle.readBuffer(chunk);
+    setImmediate(function() {
+      debug('data', chunk.length);
+      if (self._handle)
+        self._handle.readBuffer(chunk);
+    });
   });
   this.stream.once('end', function() {
-    self._handle.emitEOF();
-  });
-  this.stream.on('error', function(err) {
-    self.emit('error', err);
+    setImmediate(function() {
+      debug('end');
+      if (self._handle)
+        self._handle.emitEOF();
+    });
```

Comment:
Actually, let's drop it out, and revert it back only in case of any troubles. I no longer think that it might be reasonable. (@EricTheOne: I hope you don't mind)